### PR TITLE
Refactoring 1st PR

### PR DESCRIPTION
**Pull request by vince62s**
@srush 
FYI some fixes were commited after this first PR.
Also I had to include some others commits to resolve conflicts.

### CODE REVIEW

**Comment by srush** [comment]
Thanks! This is a great start. 

Please fix though the lint errors (described here https://github.com/OpenNMT/OpenNMT-py/blob/master/CONTRIBUTING.md ) , and I will do a review.

**Comment by vince62s** [comment]
The only remaining conflict stems from the issue with old models (img and audio from the Harvard repo) which still have the Optim class in their checkpoint.

**Comment by srush** [comment]
Hmm, so it seems like there are a couple tests still failing. Some of them are due to Optim, but there is one that just looks like it trains wrong. I'll check it out and try to fix...

For the optim issue, can we just make a pseudo-class to make this work? just make onmt.Optim point to the new path.

**Comment by srush** [comment]
Okay, sent you a PR for the Optim issue. 

Now the only test that is breaking is a translation integration test. Should there be any changes in translation? Looking into it.

**Comment by vince62s** [comment]
hmm it was passing locally, I'll check again.

**Comment by srush** [comment]
Sorry, this should be more clear. The unit tests are all passing, but travis also runs a couple simple models. You can run them with `pull_request_check.sh`. 

The one that is failing trains a small model and the runs translation. It looks like it gets a slightly worse model for some reason. 

[+] Doing NMT {preprocess + train + translation} test...18c18
< p r i l a ɡ o d ʎ i ʋ o s t
---
> p r i l a ɡ o d l i ʋ o s t
33c33
< ʎ u t ɲ a
---
> l u t n a
42c42
< s a ɡ i ɲ a t i
---
> s a ɡ i n j a t i
54c54
< b e z ʋ o ʎ a
---
> b e z ʋ o l a
91c91
< s p ʎ o ʃ t e n
---
> s p l o ʃ t e n
138c138
< n e i z ʎ e t ʃ i ʋ o s t
---
> n e i z l e t ʃ i ʋ o s t
175c175
< ʎ e t o
---
> l e t o
Failed !

**Comment by vince62s** [comment]
I don't see the "Failed!" but yes the end of theoutput is weird.
it stops at "leto"
it doesn't say what's wrong.
Any clue? the other test_model runs fine. could it be an encoding issue ?

**Comment by srush** [comment]
Yeah, not sure. I'll look a bit more.

**Comment by srush** [comment]
It looks like it is just this command that differs (no training). I'll look what it could be. 

`python translate.py -model onmt/tests/test_model2.pt  -src  data/morph/src.valid  -verbose -batch_size 10 -beam_size 10 -tgt data/morph/tgt.valid -out /tmp/trans; diff  data/morph/tgt.valid /tmp/trans`

**Comment by vince62s** [comment]
in fact the translate.py is fine. the issue comes from the "diff" command just after.
For some reason a few characters are not translated as in the gold target.
Must be some encoding or local issue.

**Comment by vince62s** [comment]
new error, but because travis failed on master

**Comment by srush** [comment]
Looks like it is an unrelated issue with a different lib, otherwise things look good. 

I'll fix master and then merge this.

**Comment by vince62s** [comment]
the issue is with joblib 0.12 in librosa. https://github.com/librosa/librosa/issues/729
maybe we can just comment the audio test until this is fixed.

File: onmt/utils/loss.py [hunk 1]
```diff
@@ -111,16 +134,15 @@ def sharded_compute_loss(self, batch, output, attns,
...
         """
-        batch_stats = onmt.Statistics()
+        batch_stats = onmt.utils.Statistics()
         range_ = (cur_trunc, cur_trunc + trunc_size)
         shard_state = self._make_shard_state(batch, output, range_, attns)
-
         for shard in shards(shard_state, shard_size):
             loss, stats = self._compute_loss(batch, **shard)
-            loss.div(float(normalization)).backward()
+            loss.div(float(normalization)).backward(retain_graph=True)
```
Comment (srush) [inline onmt/utils/loss.py]: Can you revert this file to be the same as master? It has some changes that are wrong, and we fixed in master for 0.4.

File: onmt/utils/transformer_util.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
+"""
+Implementation of "Attention is All You Need"
+"""
+
+import torch.nn as nn
+
+import onmt
+
+
+class PositionwiseFeedForward(nn.Module):
```
Comment (srush) [inline onmt/utils/transformer_util.py]: I don't think this should be a utility as it is a module.

File: pylintrc [hunk 3]
```diff
@@ -0,0 +1,426 @@
+[MASTER]
```
Comment (srush) [inline pylintrc]: Can you remove this file? We are using flake8 not pylint.

File: onmt/utils/loss.py (hunk 1 above)
Comment (srush) [reply onmt/utils/loss.py]: (For instance retain-graph actually is incorrect here and may lead to extra gradients)

File: onmt/utils/cnn_factory.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+"""
```
Comment (srush) [inline onmt/utils/cnn_factory.py]: Can you remove this from utiils as it has modules in it.

File: onmt/utils/rnn_factory.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
+"""
```
Comment (srush) [inline onmt/utils/rnn_factory.py]: This file seems unnecessarily small, can you add to the stacked module.

File: onmt/inputters/text_dataset.py [hunk 6]
```diff
@@ -216,6 +230,7 @@ def get_fields(n_src_features, n_tgt_features):
                                      pad_token=PAD_WORD)
 
         def make_src(data, vocab, is_train):
+            """ ? """
```
Comment (srush) [inline onmt/inputters/text_dataset.py]: What are these?

File: onmt/decoders/decoder.py [hunk 7]
```diff
@@ -612,13 +384,19 @@ class DecoderState(object):
...
+    #    """ Need to document this VN """
+    #    for h in self._all:
+    #        if h is not None:
+    #            h.detach_()
 
     def detach(self):
-        for h in self._all:
-            if h is not None:
-                h.detach_()
+        """ Need to document this """
```
Comment (srush) [inline onmt/decoders/decoder.py]: Is this a change in behavior? Also is this a note to us, Maybe add a "?" or a "TODO:" ?

File: onmt/utils/transformer_util.py (hunk 2 above)
Comment (vince62s) [reply onmt/utils/transformer_util.py]: yes we noticed this too.
I suggest we do this when taking care of the AAN which adds things this file.

File: pylintrc (hunk 3 above)
Comment (vince62s) [reply pylintrc]: ok you did it.
At first I didn't know about pylint vs flake so I wanted to replicate onmt-tf.
This explains also the dummy comments below in each method/function/class

File: onmt/decoders/decoder.py (hunk 7 above)
Comment (vince62s) [reply onmt/decoders/decoder.py]: @pltrdy  I don't recall exactly. This was to fix the transformer behavior while migrating to 0.4
But not only.

File: onmt/utils/transformer_util.py (hunk 2 above)
Comment (srush) [reply onmt/utils/transformer_util.py]: That's fine, although as you know, I like have PRs for reorganizing be separate from functional changes.
